### PR TITLE
feat(plugins): add Auth0 plugin

### PR DESCRIPTION
**Pull request by Shurtu-gal**
### Description
- Add Auth0 plugin for authentication

### Tasks
- [x] Make the plugin
- [x] Add README
- [x] Add tests.
- [x] https://github.com/amplication/amplication/pull/7011

### Resolves
closes https://github.com/amplication/amplication/issues/4452

### CODE REVIEW

File: plugins/auth-auth0/src/constants.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+export const clientStaticPath = join(__dirname, "static", "client");
+export const serverStaticPath = join(__dirname, "static", "server");
+export const templatesPath = join(__dirname, "templates");
+
+export const adminUIPackageJsonValues = {
+  dependencies: {
+    "@auth0/auth0-spa-js": "^2.1.2",
+    "react-router-dom": "^5.3.3",
+    history: "4.10.1",
```
Comment (overbit) [inline plugins/auth-auth0/src/constants.ts:11]: There are missing the relative @types/* devDepenencies

File: plugins/auth-auth0/src/static/client/auth-provider/ra-auth-auth0.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  domain: process.env.REACT_APP_AUTH0_DOMAIN || "",
+  clientId: process.env.REACT_APP_AUTH0_CLIENT_ID || "",
+  cacheLocation: "localstorage",
+  authorizationParams: {
+    audience: process.env.REACT_APP_AUTH0_AUDIENCE,
+    scope: "openid profile email",
+  },
+  useRefreshTokens: true,
+});
+
```
Comment (overbit) [inline plugins/auth-auth0/src/static/client/auth-provider/ra-auth-auth0.ts:16]: image

Is the login supposed to work?

**Review by overbit — CHANGES_REQUESTED** [review]

File: plugins/auth-auth0/package.json [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    "prettier": "^2.8.7",
+    "rimraf": "^4.4.1",
+    "ts-jest": "^29.1.1",
+    "ts-loader": "^9.4.2",
+    "ts-node": "^10.9.1",
+    "tsconfig-paths-webpack-plugin": "^4.1.0",
+    "typescript": "^4.9.3",
+    "webpack": "^5.76.0",
+    "webpack-cli": "^5.0.1",
+    "webpack-node-externals": "^3.0.0"
```
Comment (overbit) [inline plugins/auth-auth0/package.json:40]: Please align the dependencies with the latest master dependencies of the plugin-template

**Review by overbit — CHANGES_REQUESTED** [review]

File: plugins/auth-auth0/webpack.config.js
[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply plugins/auth-auth0/webpack.config.js:1]: Sure

File: plugins/auth-auth0/src/templates/App.template.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+//@ts-ignore
+import buildGraphQLProvider from "./data-provider/graphqlDataProvider";
+//@ts-ignore
+import { theme } from "./theme/theme";
+//@ts-ignore
+import Login from "./Login";
+import "./App.scss";
+//@ts-ignore
+import Dashboard from "./pages/Dashboard";
+import { createBrowserHistory as createHistory } from "history";
```
[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply plugins/auth-auth0/src/templates/App.template.tsx:12]: image
@types/history not needed

[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply plugins/auth-auth0/src/templates/App.template.tsx:12]: The ts-ignore was a side-effect from original template I guess.

File: plugins/auth-auth0/src/core/create-jwt-strategy.ts
[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply plugins/auth-auth0/src/core/create-jwt-strategy.ts:1]: It should be working if you are using the latest one. just tried it out.

File: plugins/auth-auth0/src/static/client/auth-provider/ra-auth-auth0.ts (hunk 2 above)
Comment (Shurtu-gal) [reply plugins/auth-auth0/src/static/client/auth-provider/ra-auth-auth0.ts:16]: Yes, its working.

File: plugins/auth-auth0/README.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+Results in configuring the app to use auth0 for authentication. It adds the necessary dependencies, creates a JWT strategy and adds the required environment variables in the `.env` file.
+
+## Usage
+
+Before using the app, make sure to have done the following :-
+
+### Create an admin user in the postgres database
+
+- Go to CustomSeed and add user details as follows ( or as per your own authEntity ) :-
```
[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply plugins/auth-auth0/README.md:110]: Any suggestions for this @mulygottlieb 🙏🏻

[the remark this answers is not in the document]
Comment (mulygottlieb) [reply plugins/auth-auth0/README.md:110]: @Shurtu-gal, I may be a bit confused but I am not really following this example:
1. You are adding below a lot of fields that are not present in the default Authentication entity that we generate ("User")
2. In particular, it seems you must have an "email" field in the authentication entity in order for this to work, but you are not specifying this in the README or anywhere else.

Can you please clarify if the user needs first to add an "email" field to the Auth entity? Should the field be named exactly "email"? (seems like it), what is the expected field type, etc. This all needs to be explained.

Then, you can give a concrete example of how the customSeed.ts file should look before/after the modifications you are suggesting here.

[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply plugins/auth-auth0/README.md:110]: Got it !!

File: plugins/auth-auth0/README.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+- Go to the settings tab and copy the values of the following :-
+
+  - Domain
+  - Client ID
+
+  ![Auth0 Application Settings](./assets/auth0-admin-settings.png)
+
+### Create an Auth0 API
+
+- Go to the dashboard and create a new API of type Machine to Machine
```
Comment (abrl91) [inline plugins/auth-auth0/README.md:39]: Did you mean application?

File: plugins/auth-auth0/README.md
Comment (abrl91) [inline plugins/auth-auth0/README.md:1]: @Shurtu-gal 
I tried to follow this readme, and I got confused.
1. Try to be more precise in the navigation in the Auth0 dashboard
2. Please change the code screenshots to code blocks. For example
```
const jwtCheck = auth({
  audience: 'https://sample-nest.demo.com',
  issuerBaseURL: 'https://dev-1jb4isqtvnywpl0o.us.auth0.com/',
  tokenSigningAlg: 'RS256'
});
```
3. There are some screenshots of code without any explanation if the user needs just to review it or copy and paste it somewhere. For example: in (2) I think you should add a few sentences explaining what you want to explain, or just remove it if it is something obvious for Atuh0 users.

File: plugins/auth-auth0/amplicationrc.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{
+  "settings": {
+    "AUTH0_DOMAIN": "AUTH0_DOMAIN",
+    "AUTH0_CLIENT_ID": "AUTH0_CLIENT_ID",
+    "AUTH0_AUDIENCE": "AUTH0_AUDIENCE",
+    "AUTH0_ISSUER_URL": "AUTH0_ISSUER_URL",
+    "recipe": {
```
Comment (overbit) [inline plugins/auth-auth0/amplicationrc.json:7]: @Shurtu-gal let's try to be consistent with the naming convention at least within the plugin itself. I'd suggest to go for camelCase all the way

**Review by overbit — CHANGES_REQUESTED** [review]

File: plugins/auth-auth0/amplicationrc.json (hunk 7 above)
Comment (Shurtu-gal) [reply plugins/auth-auth0/amplicationrc.json:7]: Sure.

**Comment by Shurtu-gal** [comment]
> @Shurtu-gal I also got this error while running the `db:init`
> 
> image
> 
> The `password.service` was not generated at all

Yeah I will be producing a new custom seed as the DSG is automatically adding password into the seed.

File: plugins/auth-auth0/README.md
Comment (Shurtu-gal) [reply plugins/auth-auth0/README.md:1]: I will improve the README and also see if I can take advantage of management API to make it smoother.

File: plugins/auth-auth0/README.md (hunk 6 above)
Comment (Shurtu-gal) [reply plugins/auth-auth0/README.md:39]: Yeah, it's an API application. I will update the README.

**Comment by Shurtu-gal** [comment]
@mulygottlieb @overbit @abrl91 All the required changes are done.

File: plugins/auth-auth0/package.json [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    "format:write": "prettier --write \"**/*.{js,ts}\"",
+    "lint:fix": "eslint './src/**/*.{ts, tsx}' --fix --c .eslintrc",
+    "lint": "eslint './src/**/*.{ts, tsx}' --c .eslintrc",
+    "test": "jest"
+  },
+  "author": "Ashish Padhy <100484401+Shurtu-gal@users.noreply.github.com>",
+  "license": "Apache-2.0",
+  "devDependencies": {
+    "@amplication/code-gen-types": "^2.0.12",
+    "@amplication/code-gen-utils": "^0.0.6",
```
Comment (overbit) [inline plugins/auth-auth0/package.json:20]: @Shurtu-gal do you mind updating these dependencies to the latest? 
plugin-template should be up-to-date

**Review by overbit — CHANGES_REQUESTED** [review]

File: plugins/auth-auth0/package.json (hunk 8 above)
Comment (Shurtu-gal) [reply plugins/auth-auth0/package.json:20]: Sure.

Comment (Shurtu-gal) [reply plugins/auth-auth0/package.json]: Done

File: plugins/auth-auth0/README.md [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+  - Domain
+  - Client ID
+
+  ![Auth0 Application Settings](./assets/auth0-admin-settings.png)
+
+### Create an Auth0 API
+
+- Go to the [API Dashboard](https://manage.auth0.com/#/apis) and create a new API with the following values :-
+
+  - Identifier : https://sample-nest.demo.com
```
Comment (overbit) [inline plugins/auth-auth0/README.md:188]:
```suggestion
  - Identifier : http://localhost:3001
```

as you show in the example here: https://github.com/amplication/plugins/pull/177/files#diff-56ba0222f9cc8ff2d6c06bcef99f053c65035b5823d83eeb60f1bf8532a608e0R200

**Review by overbit — CHANGES_REQUESTED** [review]

File: plugins/auth-auth0/README.md [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+This method uses the [Auth0 Management API](https://auth0.com/docs/api/management/v2) to get the required values.
+
+> **Note:** This method requires the user to have an auth0 account or have access to JWT token of the auth0 account.
+
+#### Method Specific Configuration
+
+This method requires the following values to be provided in the plugin settings.
+
+- `useManagementApi` : `true` ( To use this method )
```
Comment (overbit) [inline plugins/auth-auth0/README.md:99]: One note here:
The current implementation is not creating the app in the auth0 tenant in an idempotent way and if a service gets built more the one time, the plugin will throw errors like the following:
image

It's OK to open a new issue regarding this so we avoid blocking this plugin from release until it's fixed
@Shurtu-gal

**Review by overbit — CHANGES_REQUESTED** [review]

File: plugins/auth-auth0/src/static/client/auth-provider/ra-auth-auth0.ts
[the remark this answers is not in the document]
Comment (Shurtu-gal) [reply plugins/auth-auth0/src/static/client/auth-provider/ra-auth-auth0.ts:1]: @overbit This has been fixed.

File: plugins/auth-auth0/README.md (hunk 10 above)
Comment (Shurtu-gal) [reply plugins/auth-auth0/README.md:99]: [Can be seen here](https://github.com/amplication/plugins/pull/177/commits/4ba7cac451c7896d1435bff6f21180538c0d482b)

**Pull request merged**